src/rgw/rgw_sal_motr.h

```diff
@@ -40,6 +43,177 @@ class MotrStore;
+    marker.clear();
+    bl.copy(marker_len + 1, marker);
+    marker.erase(0, 1);
+
+    DECODE_START(2, bl);
+    //decode(marker, bl);
+    decode(sender, bl);
+    DECODE_FINISH(bl);
+  }
+};
+WRITE_CLASS_ENCODER(MotrWatchNotifyMsg)
+
+class MotrCacheNotif : public MotrWatchNotifyMsg {
+protected:
+  // Key of the cache item.
+  std::string key;
+  // Op to the cached item.
+  int op;
+
+public:
+  MotrCacheNotif(std::string _maker, std::string _sender, const std::string& name, int cache_op) :
+    MotrWatchNotifyMsg(_maker, _sender)
+  {
+    key.assign(name);
+    op = cache_op;
+  }
+  MotrCacheNotif(std::string _marker) : MotrWatchNotifyMsg(_marker) {}
+  MotrCacheNotif() {}
+
+  virtual void encode(bufferlist& bl) const {
+    MotrWatchNotifyMsg::encode(bl);
+
+    ENCODE_START(2, 2, bl);
+    encode(key, bl);
+    encode(op, bl);
+    ENCODE_FINISH(bl);
+  };
+
+  virtual void decode(bufferlist::const_iterator& bl) {
+    MotrWatchNotifyMsg::decode(bl);
+
+    DECODE_START(2, bl);
+    decode(key, bl);
+    decode(op, bl);
+    DECODE_FINISH(bl);
+  }
+
+  std::string& get_key() { return key; }
+  int get_op() { return op; }
+  std::string& get_notifier() { return this->sender; }
+};
+WRITE_CLASS_ENCODER(MotrCacheNotif)
+
+// MotrWatcher registers a FDMI filter and a FDMI callback function.
+// When FDMI records (picked up by the filter) arrive, the callback function
+// process the records and retrieve the information sent by the notifier.
+class MotrWatcher {
+protected:
+  CephContext *cctx{nullptr};
+  const struct m0_fdmi_pd_ops *fdmi_dock_ops{nullptr};
+  struct m0_fid fdmi_plugin_fid;
+  struct m0_fdmi_plugin_ops fdmi_plugin_cb;
+
+  // The list of notifiers that it doesn't want to receive notifications
+  // from. For example, a metadata cache has a pair of notifier and watcher.
+  // The watcher has to exclude the notifications sent by its friend notifier.
+  std::list<std::string> excluded_notifiers;
+
+public:
+  MotrWatcher(CephContext *_cctx) : cctx(_cctx) {}
+  MotrWatcher() {}
+  virtual ~MotrWatcher() {}
+  // Callback function of the watcher's user. For example, if watcher is
+  // user by object metadata cache, this callback function is called to
+  // trigger corresponding cache actions.
+  virtual int watch_cb(bufferlist& bl) = 0;
+  int init_fdmi_plugin(const DoutPrefixProvider *dpp);
+
+  void exclude_notifier(std::string notifier) {
+    excluded_notifiers.push_back(notifier);
+  }
+
+  bool is_excluded_notifier(std::string& notifier) {
+    auto iter = std::find(excluded_notifiers.begin(), excluded_notifiers.end(), notifier);
+    return iter == excluded_notifiers.end()? false : true;
+  }
+};
+
+class MotrMetaCache;
+class MotrCacheWatcher : public MotrWatcher {
+protected:
+  MotrMetaCache *cache;
+
+public:
+  MotrCacheWatcher(CephContext *_cctx, MotrMetaCache* _cache) : MotrWatcher(_cctx), cache(_cache){}
+  virtual int watch_cb(bufferlist& bl) override;
+};
+
+class MotrStore;
+class MotrNotifier {
```

Comment:
Add some comments about the purpose of this class.

Reply:
will add